pkg/liqo-controller-manager/virtualNode-controller/virtualnode_controller.go

```diff
@@ -0,0 +1,113 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package virtualnodectrl
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+)
+
+const (
+	virtualNodeControllerFinalizer = "virtualnode-controller.liqo.io/finalizer"
+)
+
+// VirtualNodeReconciler manage NamespaceMap lifecycle.
+type VirtualNodeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes/status,verbs=get;update;patch
+
+// Reconcile manage NamespaceMaps associated with the virtual-node.
+func (r *VirtualNodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	virtualNode := &corev1.Node{}
+	if err := r.Get(ctx, req.NamespacedName, virtualNode); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("There is no a virtual-node called '%s'", req.Name)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf(" %s --> Unable to get the virtual-node '%s'", err, req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// If the deletion timestamp is set all the NamespaceMaps associated with the virtual-node must be deleted.
+	if !virtualNode.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, r.removeAssociatedNamespaceMaps(ctx, virtualNode)
+	}
+
+	// The virtual-node must have the cluster-id annotation
+	if _, ok := virtualNode.GetAnnotations()[liqoconst.RemoteClusterID]; !ok {
+		err := fmt.Errorf("the annotation '%s' is not found on node '%s'", liqoconst.RemoteClusterID, virtualNode.GetName())
+		klog.Error(err)
+		return ctrl.Result{}, err
+	}
```

Comment:
Should be moved above the previous function, as the annotation is used also in that case.

Reply:
you are right, thank you